core/src/autogluon/core/hpo/executors.py

```diff
@@ -0,0 +1,370 @@
+    
+    def initialize(self, hyperparameter_tune_kwargs, default_num_trials=None, time_limit=None):
+        self.time_limit = time_limit
+        if isinstance(hyperparameter_tune_kwargs, dict):
+            hyperparameter_tune_kwargs = hyperparameter_tune_kwargs.copy()
+        if isinstance(hyperparameter_tune_kwargs, str):
+            hyperparameter_tune_kwargs = self.custom_to_ray_preset_map[hyperparameter_tune_kwargs]
+        hyperparameter_tune_kwargs['scheduler'] = self.custom_to_ray_scheduler_preset_map.get(
+            hyperparameter_tune_kwargs['scheduler'],
+            hyperparameter_tune_kwargs['scheduler']
+        )
+        hyperparameter_tune_kwargs['searcher'] = self.custom_to_ray_searcher_preset_map.get(
+            hyperparameter_tune_kwargs['searcher'],
+            hyperparameter_tune_kwargs['searcher']
+        )
+        if 'num_trials' not in hyperparameter_tune_kwargs and default_num_trials is not None:
+            hyperparameter_tune_kwargs['num_trials'] = default_num_trials
+        self.hyperparameter_tune_kwargs = hyperparameter_tune_kwargs
+        return time_limit
+    
+    def register_resources(self, initialized_model_base):
+        user_cpu_count = initialized_model_base._get_child_aux_val(key='num_cpus', default=None)
+        user_gpu_count = initialized_model_base._get_child_aux_val(key='num_gpus', default=None)
+        resource_kwargs = initialized_model_base._preprocess_fit_resources()
+
+        num_gpus = ResourceCalculator.get_total_gpu_count(
+            user_specified_num_gpus=user_gpu_count,
+            model_default_num_gpus=resource_kwargs.get('num_gpus', 0),
+        )
+        num_cpus = ResourceCalculator.get_total_cpu_count(
+            user_specified_num_cpus=user_cpu_count,
+            model_default_num_cpus=resource_kwargs.get('num_cpus', 0),
+        )
+        
+        self.resources = dict(num_gpus=num_gpus, num_cpus=num_cpus)
+        
+    def validate_search_space(self, search_space, model_name):
+        from ray.tune.sample import Domain
+        if not any(isinstance(search_space[hyperparam], (Space, Domain)) for hyperparam in search_space):
+            logger.warning(f"\tNo hyperparameter search space specified for {model_name}. Skipping HPO. "
+                           f"Will train one model based on the provided hyperparameters.")
+            raise EmptySearchSpace
+        self.search_space = search_space
+        logger.log(15, f"\tHyperparameter search space for {model_name}: ")
+        for hyperparam in search_space:
+                if isinstance(search_space[hyperparam], (Space, Domain)):
+                    logger.log(15, f"{hyperparam}:   {search_space[hyperparam]}")
+                    
+    def execute(
+            self,
+            model_trial,
+            train_fn_kwargs,
+            directory,
+            minimum_cpu_per_trial,
+            minimum_gpu_per_trial,
+            model_estimate_memory_usage,
+            adapter_type,
```

Comment:
add type hints if possible, will make it easier to understand given there are no default values provided.

Also consider making this not work with positional arguments, very easy to make a mistake here.

Reply:
Probably want to add docs here as well